include/indicators/progress_bar.hpp

```diff
@@ -34,80 +34,95 @@ SOFTWARE.
 #include <chrono>
 #include <cmath>
 #include <indicators/color.hpp>
+#include <indicators/setting.hpp>
+#include <tuple>
 #include <iomanip>
 #include <iostream>
 #include <mutex>
 #include <string>
 #include <thread>
+#include <tuple>
+#include <type_traits>
 
 namespace indicators {
 
 class ProgressBar {
+  using Settings = std::tuple<
+      option::BarWidth,
+      option::PrefixText,
+      option::PostfixText,
+      option::Start,
+      option::End,
+      option::Fill,
+      option::Lead,
+      option::Remainder,
+      option::MaxPostfixTextLen,
+      option::Completed,
+      option::ShowPercentage,
+      option::ShowElapsedTime,
+      option::ShowRemainingTime,
+      option::SavedStartTime,
+      option::ForegroundColor
+  >;
 public:
-  void set_foreground_color(Color color) {
-    std::lock_guard<std::mutex> lock{_mutex};
-    _foreground_color = color;
+  template <typename... Args, typename std::enable_if<details::are_settings_from_tuple<Settings, typename std::decay<Args>::type...>::value, void*>::type = nullptr>
+  explicit ProgressBar(Args&&... args) :
+    settings_(
+        details::get<details::ProgressBarOption::bar_width>(option::BarWidth{100}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::prefix_text>(option::PrefixText{}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::postfix_text>(option::PostfixText{}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::start>(option::Start{"["}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::end>(option::End{"]"}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::fill>(option::Fill{"="}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::lead>(option::Lead{">"}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::remainder>(option::Remainder{" "}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::max_postfix_text_len>(option::MaxPostfixTextLen{0}, std::forward<Args>(args)...),
+        details::get<details::ProgressBarOption::completed>(option::Completed{false}, std::forward<Args>(args)...),
```

Comment:
I'm questioning whether `Completed` should be a setting. Although it can be set by the user, It's more of a state than a configurable setting. Might be better to have this be a separate method `bar.mark_as_complete()` (or something like it) instead of `bar.set_option(option::Completed{true})`. Not a huge issue for me, just thinking out loud.

Reply:
Correct I agree.

This is the reason why I did not remove the _mark_as_complete_ function. So right now there are two ways to do that - by __set_option__ and by __mark_as_complete__.

The question is, whether or not you would like to remove the possibility of setting the Completion state via set_option.

Reply:
I think we can keep it. If we choose to call anything parametric in the class as a `Setting/Option`, then `Completed` qualifies. I don't mind it since we also have `mark_as_complete`.